src/gpuarray_blas_cuda_cublas.c

```diff
@@ -171,28 +240,53 @@ static int sgemmBatch(cb_order order, cb_transpose transA, cb_transpose transB,
                       gpudata **B, size_t *offB, size_t ldb,
                       float beta, gpudata **C, size_t *offC, size_t ldc,
                       size_t batchCount) {
-  FETCH_CONTEXT(A[0]);
-  FUNC_DECLS;
-  PREP_ORDER_GEMMBATCH;
+  cuda_context *ctx;
+  size_t *lt, t;
+  gpudata **T;
+  size_t i;
+  cb_transpose transT;
+  cublasStatus_t err;
+
+  if (batchCount == 0) return GA_NO_ERROR;
+
+  ctx = A[0]->ctx;
+
+  /* Possibly optimize this to make multiple dispatch of sgemm for
+   * bigger sizes */
   float **T_l = alloca(sizeof(float *) * batchCount * 3);
   const float **A_l = (const float **)T_l;
   const float **B_l = (const float **)T_l + batchCount;
   float **C_l = T_l + (batchCount * 2);
   CUdeviceptr Ta, Aa, Ba, Ca;
 
-  HANDLE_ORDER_GEMMBATCH;
-  FUNC_INIT;
-
-  {
-    size_t i;
-    for (i = 0; i < batchCount; i++) {
-      ARRAY_INIT(A[i]);
-      A_l[i] = ((float *)A[i]->ptr) + offA[i];
-      ARRAY_INIT(B[i]);
-      B_l[i] = ((float *)B[i]->ptr) + offB[i];
-      ARRAY_INIT(C[i]);
-      C_l[i] = ((float *)C[i]->ptr) + offC[i];
-    }
+  if (order == cb_c) {
+    /* swap A and B */
+    t = N;
+    N = M;
+    M = t;
+    T = A;
+    A = B;
+    B = T;
+    t = lda;
+    lda = ldb;
+    ldb = t;
+    transT = transA;
+    transA = transB;
+    transB = transT;
+    lt = offA;
+    offA = offB;
+    offB = lt;
+  }
+
+  cuda_enter(ctx);
+
+  for (i = 0; i < batchCount; i++) {
+    cuda_wait(A[i], CUDA_WAIT_READ);
+    cuda_wait(B[i], CUDA_WAIT_READ);
+    cuda_wait(C[i], CUDA_WAIT_READ|CUDA_WAIT_WRITE);
```

Comment:
It may be useful, but would be too complex to communicate in a non-annoying way through the interface.  Since I don't see any calls anywhere to this function I also fail to see any use case for now.

If we have a use case where A, B and C are essentially a 3D tensors that we perform gemm elementwise on, a different interface would be much faster.  Otherwise I can look at the problem and try to come up with a reasonable interface if possible.